xarray/core/dataarray.py

```diff
@@ -279,6 +302,7 @@ def __init__(
                 encoding = getattr(data, 'encoding', None)
 
             data = as_compatible_data(data)
+            data = _check_data_shape(data, coords, dims)
```

Comment:
I wonder if we should move this logic *above* `as_compatible_data`, which would let us distinguish between scalar values like `float`/`int` (which don't have an inherent `shape`) vs 0-dimensional NumPy arrays (which *do* have an array shape already).

For example:
- `xarray.DataArray(0.5, coords=[('x', np.arange(3)), ('y', ['a', 'b'])])` -> duplicate the scalar to make an array of shape `(3, 2)`
- `xarray.DataArray(np.array(1.0), coords=[('x', np.arange(3)), ('y', ['a', 'b'])])` -> error, shapes do not match

Reply:
If I understand correctly, the second example you provided _shouldn't_ work since `np.array(1.0)` is a 0-dimensional NumPy array with shape `()` and `DataArray` expects it to have a `(3, 2)` shape, right?. The current behavior is set to duplicate the value as if it were `xarray.DataArray(1.0, coords=[('x', np.arange(3)), ('y', ['a', 'b'])])`, which I thought was the desired feature. I am currently pushing a commit that makes this work since I didn't consider the case of `coords` being a list of tuples (although all test passed).

Regarding the `_check_data_shape` position, I placed it after `as_compatible_data` since the latter returns an `ndarray` containing the value passed to it, scalar or `None`, on which I can check the shape.